### PR TITLE
Properly obtain a tenant ID for an event and improve `BlackBoxBoundedContext`

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR fixes issues with entities state verification using `BlackBoxBoundedContext`.

Previously, the state verification didn't consider a `TenanId` during the querying of entities. Also, `BlackBoxBoundedContext` always was multitenant. To make things clearer, the bounded context was split into the `MultitenantBlackBoxContext` and `SingleTenantBlackBoxContext`.

The only difference of a `MultitenantBlackBoxContext` is that it requires to specify a tenant ID. Then, the posting of messages and verifications are performed within the tenant.

Other changes in the `BlackBoxBoundedContext` API:
- Added `receivesEventsProducedBy` method to easily set a producer of received events.
- Added ability to set events `Enricher` (to test enrichments).

Also, during the fix, the related issue  #886 was found. The issue was in the wrong obtaining of a tenant ID in several places. To fix the problem, `EventEnvelope.getActorContext()` was changed and now it considers an import context and nested event contexts.

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyState.java [hunk 1]
```diff
@@ -98,4 +132,10 @@ protected void compare(ImmutableCollection<? extends Message> expected,
             }
         };
     }
+
+    /**
+     * Produces {@link VerifyState} based on a {@link TenantId}.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyState.java]: Please expand this description.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyState.java [hunk 2]
```diff
@@ -77,19 +81,49 @@ protected abstract void compare(ImmutableCollection<? extends Message> expected,
                                     ImmutableCollection<? extends Message> actual);
 
     /**
-     * Obtains a verifier which checks that the system contains exactly the passed entity states.
+     * Obtains the producer of {@link #exactly(TenantId, Class, Iterable)}.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyState.java]: In this changeset you mix the verb `produce` with the term `producer` (the latter meaning an entity which is responsible for emitting an event or a command). I don't think it's fine.

In this particular piece of the documentation the whole text looks misleading. I.e. it's confusing for me as a user why I would want to use this method anyway. Besides, the `@return` section tells lies.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/SingletenantBlackBoxContext.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import io.spine.server.event.Enricher;
+import io.spine.testing.client.TestActorRequestFactory;
+
+import java.util.List;
+
+/**
+ * A black box bounded context for writing integration tests in a single tenant environment.
+ */
+@VisibleForTesting
+public class SingletenantBlackBoxContext
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/SingletenantBlackBoxContext.java]: It's `SingleTenantBlackBoxContext`.

File: server/src/test/java/io/spine/server/aggregate/AggregateRepositoryTest.java [hunk 4]
```diff
@@ -499,7 +499,7 @@ void afterCommand() {
                     .newBuilder()
                     .setProjectId(id)
                     .build();
-            BlackBoxBoundedContext.newInstance()
+            BlackBoxBoundedContext.singletenant()
```
Comment (armiol) [inline server/src/test/java/io/spine/server/aggregate/AggregateRepositoryTest.java]: Can we please have `newInstance()` back, meaning "single-tenant context" by default.

File: server/src/test/java/io/spine/server/aggregate/ApplyAllowImportTest.java [hunk 5]
```diff
@@ -48,13 +49,12 @@
 @DisplayName("Aggregate which supports event import should")
 class ApplyAllowImportTest {
 
-    private BlackBoxBoundedContext boundedContext;
+    private SingletenantBlackBoxContext boundedContext;
 
     @BeforeEach
     void setUp() {
-        boundedContext = BlackBoxBoundedContext
-                .newInstance()
```
Comment (armiol) [inline server/src/test/java/io/spine/server/aggregate/ApplyAllowImportTest.java]: There was a reason to format the expression like that.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import java.util.Collection;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Lists.asList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * An input of a {@link BlackBoxBoundedContext}, which receives domain messages.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java]: No idea how an "input" can receive anything. Please think of a better phrasing.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        this.commandBus = boundedContext.getCommandBus();
+        this.eventBus = boundedContext.getEventBus();
+        this.importBus = boundedContext.getImportBus();
+        this.requestFactory = checkNotNull(requestFactory);
+        this.eventFactory = eventFactory(requestFactory);
+        this.observer = checkNotNull(observer);
+    }
+
+    /**
+     * Sends off a provided command to the Bounded Context.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java]: Well, it is weird. The method is `receivesCommands`, but you say `sends`.

Also, a method is about multiple commands.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+     *
+     * @param domainEvents
+     *         a list of domain event to be dispatched to the Bounded Context
+     */
+    void receivesEvents(Collection<Message> domainEvents) {
+        List<Event> events = toEvents(domainEvents, eventFactory);
+        eventBus.post(events, observer);
+    }
+
+    void receivesEvents(Object producerId, EventMessage firstEvent, EventMessage... otherEvents) {
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/BlackBoxInput.java]: And these methods aren't documented. For some reason.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/MultitenantBlackBoxContext.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     * @param tenantId
+     *         an identifier of a tenant that is executing requests in this Bounded Context
+     * @return a new request factory instance
+     */
+    private static TestActorRequestFactory requestFactory(TenantId tenantId) {
+        return TestActorRequestFactory.newInstance(MultitenantBlackBoxContext.class, tenantId);
+    }
+
+    private static class IsTenantCommand implements Predicate<Command> {
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/MultitenantBlackBoxContext.java]: Please document. And I think when you do that, you will have to rename this class to match the description.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyMessages.java [hunk 10]
```diff
@@ -26,7 +26,7 @@
  * @param <E> the type of emitted messages
  * @author Alexander Yevsyukov
  */
-public interface Verify<E extends EmittedMessages> {
+public interface VerifyMessages<E extends EmittedMessages> {
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/VerifyMessages.java]: Not sure why this got renamed. It was named `Verify` to make the code usages to read fluently, i.e. "to verify<what>".

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/MultitenantBlackBoxContext.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+     * @return a new request factory instance
+     */
+    private static TestActorRequestFactory requestFactory(TenantId tenantId) {
+        return TestActorRequestFactory.newInstance(MultitenantBlackBoxContext.class, tenantId);
+    }
+
+    /**
+     * A predicate to match commands with a {@linkplain #tenantToMatch tenant ID}.
+     */
+    private static class HasTenantId implements Predicate<Command> {
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/MultitenantBlackBoxContext.java]: I think it is `MatchesTenant`.

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/verify/package-info.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This package provides classes for verification of a server-side of a black box bounded context.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/verify/package-info.java]: This is weird. To me it's always the "server-side" in our case. Maybe you meant "write-side" or "read-side"?

File: testutil-server/src/main/java/io/spine/testing/server/blackbox/verify/state/VerifyState.java [hunk 13]
```diff
@@ -18,84 +18,89 @@
...
-    private final ImmutableCollection<? extends Message> expectedResult;
-
-    VerifyState(Query query, ImmutableCollection<? extends Message> result) {
-        this.query = query;
-        expectedResult = result;
-    }
-
     /**
-     * Verifies the entity states.
+     * Queries and verifies entity states.
```
Comment (armiol) [inline testutil-server/src/main/java/io/spine/testing/server/blackbox/verify/state/VerifyState.java]: I think it's `Queries the read-side of the provided {@code BoundedContext} and {@linkplain here.goes.the.method.reference verifies} the results according to the pre-set expectations`.

File: testutil-server/src/test/java/io/spine/testing/server/blackbox/BlackBoxBoundedContextTest.java [hunk 14]
```diff
@@ -42,30 +46,96 @@
...
+                                      new BbProjectViewRepository());
     }
 
     @AfterEach
     void tearDown() {
         projects.close();
     }
 
+    /**
+     * Creates a new instance of a bounded context for the basic set of tests.
```
Comment (armiol) [inline testutil-server/src/test/java/io/spine/testing/server/blackbox/BlackBoxBoundedContextTest.java]: I am always confused by the "basic" term. Is there also an intermediate set of tests or an advanced one?

**Pull request merged**